openjdk/mmtkUpcalls.cpp

```diff
@@ -190,27 +190,18 @@ struct MaybeUninit {
   char _data[sizeof(T)];
 };
 
-static MaybeUninit<JavaThreadIteratorWithHandle> jtiwh;
-static bool mutator_iteration_start = true;
+static void mmtk_new_java_thread_iterator(void* iter) {
+  *(JavaThreadIteratorWithHandle*)iter = JavaThreadIteratorWithHandle();
```

Comment:
This step moves the constructed `JavaThreadIteratorWithHandle` instance to another location.

However, part of the race-prevention algorithm in OpenJDK includes letting each `Thread` hold a pointer to the `SafeThreadsListPtr` instance inside `JavaThreadIteratorWithHandle`.  If the `SafeThreadsListPtr` is moved, the field in `Thread` becomes a dangling pointer.

See

-   https://github.com/mmtk/openjdk/blob/e1e420b3c4fd627218e45ec977972561c559acec/src/hotspot/share/runtime/threadSMR.cpp#LL381C13-L381C13
-   https://github.com/mmtk/openjdk/blob/e1e420b3c4fd627218e45ec977972561c559acec/src/hotspot/share/runtime/threadSMR.cpp#L416